### DISCUSSION TITLE
Access to API endpoints for functionality available within app

### DISCUSSION THREAD

**Discussion by IPertinax** [amberelectric/public-api#230] [General]
Is there another level of access to API that can be permissioned for user consumption?  The default [developer API](https://app.amber.com.au/developers/) is really quite featureless in that it provides simple read pricing information and not functionality to control or interact though Smartshift to inverters.  I do not see using HA to tell me of a price spike as particularly useful (to me) when I could be tooling to control the overall managment of my inverter/battery without taking it out of Amber completely.

Similarly, the Amber app that we can interact with is both low in quality and usability. Obviously, the app is talking to an API so the endpoints are there, they just have not been exposed to public for consumption and control of own destiny. 

Is there a feature rich API for public consumption and if not are there plans to provide this?

**Comment by madpilot** [comment]
> Is there a feature rich API for public consumption and if not are there plans to provide this?

We're working on a Smartshift API at the moment, though no concrete ETA as there a bunch of architectural changes that need to happen before it can be released.

It will allow you to fetch the current telemetry, list of events you have sent, and send self-consume, charge, discharge, preserve-charge and curtailment events.

> Obviously, the app is talking to an API so the endpoints are there, they just have not been exposed to public for consumption and control of own destiny.

These APIs are private, and not supported for out-of-app use.

**Comment by IPertinax** [comment]
Thanks for the response Myles.

 

There doesn’t appear to be an app roadmap, so when you say no concrete ETA are we talking months or years?

**Reply by madpilot** [reply]
I don't know - hence "no ETA". It's blocked by other work at the moment which is not scheduled for the next quarter, and we tend to no put timelines on things further away then that. Next year, probably?

**Comment by justinwunda** [comment]
I'd just like to also register my interest for the Smartshift API in the probably vain hope that it might get bumped up on a priorities list somewhere >:D I agree with the OP that the current state of SmartShift functionality is chafing me a bit and more granular control will be much appreciated when it comes. Thanks!

**Comment by IPertinax** [comment]
Exposing the simple functionality to disable/enable Smartshift would be a great start.  The API endpoint already exists as is available through app, as must the authentication and account/tenant corral functionality, so it should not be too greater leap to expose.

Without this ability, it really means we cannot take control from time to time programmatically as Smartshift overrides any config changes we would make directly to invertor.  Toggling Smartshit manually when we want to override programmatically removes any real possibility of augmented automation.

**Comment by Mantrid42** [comment]
I'm hanging out for this ability too. I managed to fall asleep until just after 3pm today, and due to the cloudy weather my battery didn't have enough charge to make it until 9pm. I'd love to use HA to set up automations to avoid that kind of thing.

If you need any beta testers.. ;)

**Reply by dpb03** [reply]
Count me in as a beta tester as well!

While the other functionality would be great, I'm by far most interested in "send self-consume, charge, discharge, preserve-charge". This is already in the app (so perhaps easiest to add?) and would be so useful.

**Comment by Iwanttobesomething** [comment]
I am keen to be able to control my battery and inverter based on amber prices through HA as well.

Any Updates?

**Reply by rcruikshank** [reply]
You should look into EMHASS. 
Here's a link to my doco on my system. 
Reason I set up EMHASS is I have a Sonnen battery which does not permit amber access to their VPP API so I have to manage my own battery. Luckily Sonnen hardware has a very capable RESTful API for full control by the owner. 

https://cruikshank-my.sharepoint.com/:f:/g/personal/robert_cruikshank_onmicrosoft_com/EliM7mPksJpBtEaYwf7q920BxHMk4ur2aXqf_6aQB5AO1A?e=5zmaVB

**Reply by mbourne** [reply]
@rcruikshank Your link returned a 404.

**Reply by rcruikshank** [reply]
New link

https://cruikshank-my.sharepoint.com/:f:/g/personal/robert_cruikshank_xyz/IgBYjO5j5LCaQbRGmMH-6vdtAcRzJOLq9ml6n_-mkAeQDtQ?e=si60tM

**Comment by smarmau** [comment]
Also would love to see this.  

Currently writing an app to integrate amber with my battery/solar. Moved from a Tesla to a porsche which isn't currently Amber comptabile.  Will control via a few basic rules via the porsche connect API but would really benefit from the additional data provided

**Comment by purcell-lab** [comment]
#165

**Comment by Lindsay-Mathieson** [comment]
Add me to the list, I'd love to do my own API Automation of charge/discharge events, SmartShift is not doing it all well for me. I have a much better idea of my own needs.

**Comment by bolagnaise** [comment]
Add me to the list.  SS can be incredibly frustrating when it doesn't react the way you expect it too. I would love to be able to issue charge and discharge commands through AMBER, especially with a PW3 which doesn't allow discharge commands.

**Comment by abraham-chan** [comment]
Hi @madpilot, you mentioned earlier your timelines are quarterly. Now that we're in Q3, any updates? 🤞

**Comment by hgabha** [comment]
Keen to hear about this as well. As I would like to also interact with smartshift or just schedule my own events like charge during the night when I know next day its going to rain and stay cloudy etc.
I think this next level API control will make Amber much more attractive.

**Comment by repomonkey** [comment]
Any chance this will happen this century?

**Reply by Lindsay-Mathieson** [reply]
> Any chance this will happen this century?

If its anything like their support, then no :(

**Reply by madpilot** [reply]
It's not going to happen until we stabilise the systems that is what is, ironically given the comment, killing customer support.

I know you want it, but our priority is making sure everything is scaling before we unleash a bunch of API calls at the system.

**Reply by madpilot** [reply]
> > Any chance this will happen this century?
> 
> 
> 
> If its anything like their support, then no :(

Our support wait times suck, we know. It sucks, we hate it. I get you need to vent, but these comments aren't useful. Go hang out on Facebook is you want to bitch about it.

**Reply by Lindsay-Mathieson** [reply]
> Our support wait times suck, we know. It sucks, we hate it. I get you need to vent, but these comments aren't useful. Go hang out on Facebook is you want to bitch about it.

My apologies, I can appreciate you guys are stuck between a rock and hard place, probably short on resources as expert cloud talent is hard to find.

**Reply by Pseudophryne** [reply]
> It's not going to happen until we stabilise the systems that is what is, ironically given the comment, killing customer support.
> 
> I know you want it, but our priority is making sure everything is scaling before we unleash a bunch of API calls at the system.

Hey Myles, I get where you're coming from - sounds a lot like my world 😃 There's 2 sides of this conversation.
1. Having ability to read what Smartshift is planning
2. Having ability to interact/control Smartshift

Being part of a VPP we give the control to Amber (we agreed that on sign-up) so point 2 is much less important (others will disagree).  Personally, I'm perfectly OK with Amber controlling the battery I'd just like to know what and when the VPP is planning so I can adjust everything else I have the ability to adjust at home e.g. EV charging, Hot Water, Dishwasher, etc.

**Reply by madpilot** [reply]
Yeah, I appreciate the clarification, and a read only is simpler.

I still can't get a heap of developer time on it for the moment as we are focussing on reducing the load on customer support.

**Comment by zemerdon** [comment]


**Reply by bolagnaise** [reply]
I made something for you then https://github.com/bolagnaise/PowerSync

**Reply by axilotl** [reply]
Wow. This looks fantastic. Makes me wish I had seen it before I spent a week configurating HAEO to simulate my to-be topology in readiness for hardware installation. TBH, my only roadblock right now for optimisation is getting solar forecast data for feed in tarrifs.

But I'll probably take a squizz anyway because it looks cool.

**Reply by IPertinax** [reply]
Take a look at https://api.forecast.solar

**Reply by axilotl** [reply]
@bolagnaise you're a champion. I got my solar installed and spent over a week tweaking dials and knobs, creating custom automations, scripts and sensors to get HAEO to work as intelligently as I wanted it to. 

Got 90% of the way there but the remaining 10% was a show stopper for me - ie: **stop** importing during demand windows and **smarter** exploitation of dirt cheap import prices. I finally gave up today, installed Powersync and did some initial analysis on it's recommendations. 

It seems to work _exactly_ how I wanted HAEO to work. You've even got demand window tracking in there.  If it pans out, it'll make 95% of my custom code redundant! 

Cheers for the link  :+1:
